spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/dsl/Tokens.java

```diff
@@ -0,0 +1,244 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.data.core.dsl;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Class that converts an expression into a list of {@link Token tokens}.
+ * Furthermore, this class provides methods to process the tokens and
+ * keeps track of the current token being processed.
+ *
+ * @author Andy Clement
+ * @author Patrick Peralta
+ */
+public class Tokens {
+
+	/**
+	 * Expression string to be parsed.
+	 */
+	private final String expression;
+
+	/**
+	 * List of tokens created from {@link #expression}.
+	 */
+	private final List<Token> tokenStream;
+
+	/**
+	 * Index of stream token currently being processed.
+	 */
+	private int position = 0;
+
+	/**
+	 * Index of last stream token that was successfully processed.
+	 */
+	private int lastGoodPosition = 0;
+
+
+	/**
+	 * Construct a {@code TokenProcessor} based on the provided string expression.
+	 *
+	 * @param expression string expression to convert into {@link Token tokens}.
+	 */
+	public Tokens(String expression) {
+		this.expression = expression;
+		this.tokenStream = Collections.unmodifiableList(new Tokenizer(expression).getTokens());
+	}
+
+	/**
+	 * Return the expression string converted to tokens.
+	 *
+	 * @return expression string
+	 */
+	protected String getExpression() {
+		return expression;
+	}
+
+	/**
+	 * Decrement the current token position and return the new position.
+	 *
+	 * @return new token position
+	 */
+	protected int decrementPosition() {
+		return --position;
+	}
+
+	/**
+	 * Return the current token position.
+	 *
+	 * @return current token position
+	 */
+	protected int position() {
+		return position;
+	}
+
+	/**
+	 * Return an immutable list of {@link Token tokens}
+	 *
+	 * @return list of tokens
+	 */
+	public List<Token> getTokenStream() {
+		return tokenStream;
+	}
+
+	/**
+	 * Return {@code true} if the token in the position indicated
+	 * by {@link #position} + {@code distance} matches
+	 * the token indicated by {@code desiredTokenKind}.
+	 *
+	 * @param distance number of token positions past the current position
+	 * @param desiredTokenKind the token to check for
+	 * @return true if the token at the indicated position matches
+	 * {@code desiredTokenKind}
+	 */
+	protected boolean lookAhead(int distance, TokenKind desiredTokenKind) {
+		if ((position + distance) >= tokenStream.size()) {
+			return false;
+		}
+		Token t = tokenStream.get(position + distance);
+		return t.kind == desiredTokenKind;
+	}
+
+	/**
+	 * Return {@code true} if there are more tokens to process.
+	 *
+	 * @return {@code true} if there are more tokens to process
+	 */
+	protected boolean more() {
```

Comment:
maybe change to `hasNext()`? (or at least `hasMore()` but the latter is more consistent with an Iterator)